synapse/logging/context.py

```diff
@@ -221,14 +220,13 @@ class ContextRequest:
 class _Sentinel:
     """Sentinel to represent the root context"""
 
-    __slots__ = ["previous_context", "finished", "request", "scope", "tag"]
+    __slots__ = ["previous_context", "finished", "request", "tag"]
```

Comment:
We don't need the `opentracing` `scope` stuff in here anymore.

The `scopecontextmanager.py` is no more (doesn't seem necessary)

Reply:
Maybe we do need it, seems to be the source of all these `Failed to detach context` errors from OTEL.

Related code [`synapse/metrics/background_process_metrics.py#L233-L240`](https://github.com/matrix-org/synapse/blob/1c910e2216a2f7fa57971ffbdede509fd493ebb0/synapse/metrics/background_process_metrics.py#L233-L240)
```
2022-08-02 15:23:51,469 - opentelemetry.context - 155 - ERROR - resume_sync_partial_state_room-0 - Failed to detach context
Traceback (most recent call last):
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1660, in _inlineCallbacks
    result = current_context.run(gen.send, result)
StopIteration: []

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/opentelemetry/context/__init__.py", line 153, in detach
    _RUNTIME_CONTEXT.detach(token)  # type: ignore
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/opentelemetry/context/contextvars_context.py", line 56, in detach
    self._current_context.reset(token)  # type: ignore
ValueError: <Token var=<ContextVar name='current_context' default={} at 0x111b80cc0> at 0x111c7c5c0> was created in a different Context
```


```
2022-08-02 15:23:55,137 - opentelemetry.context - 155 - ERROR - GET-8 - Failed to detach context
Traceback (most recent call last):
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1660, in _inlineCallbacks
    result = current_context.run(gen.send, result)
StopIteration: ['signed_curve25519']

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1660, in _inlineCallbacks
    result = current_context.run(gen.send, result)
StopIteration: ['signed_curve25519']

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1660, in _inlineCallbacks
    result = current_context.run(gen.send, result)
StopIteration: SyncResult(next_batch=StreamToken(room_key=RoomStreamToken(topological=None, stream=1787, instance_map=frozendict.frozendict({})), presence_key=484120, typing_key=0, receipt_key=341, account_data_key=1242, push_rules_key=8, to_device_key=58, device_list_key=199, groups_key=0), presence=[UserPresenceState(user_id='@ericgittertester:my.synapse.server', state='online', last_active_ts=1659471835107, last_federation_update_ts=1659471458095, last_user_sync_ts=1659471835107, status_msg=None, currently_active=True)], account_data=[], joined=[], invited=[], knocked=[], archived=[], to_device=[], device_lists=DeviceListUpdates(changed=set(), left=set()), device_one_time_keys_count={'signed_curve25519': 50}, device_unused_fallback_key_types=['signed_curve25519'])

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/opentelemetry/context/__init__.py", line 153, in detach
    _RUNTIME_CONTEXT.detach(token)  # type: ignore
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/opentelemetry/context/contextvars_context.py", line 56, in detach
    self._current_context.reset(token)  # type: ignore
ValueError: <Token var=<ContextVar name='current_context' default={} at 0x111b80cc0> at 0x111fa6440> was created in a different Context
2022-08-02 15:23:55,138 - opentelemetry.context - 155 - ERROR - GET-8 - Failed to detach context
Traceback (most recent call last):
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/opentelemetry/context/__init__.py", line 153, in detach
    _RUNTIME_CONTEXT.detach(token)  # type: ignore
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/opentelemetry/context/contextvars_context.py", line 56, in detach
    self._current_context.reset(token)  # type: ignore
ValueError: <Token var=<ContextVar name='current_context' default={} at 0x111b80cc0> at 0x111fe0880> was created in a different Context

```

Reply:
Seems to be a problem with Twisted deffered not supporting `contextvars`/`ContextVar` stuff.

I noticed that @richvdh made a fix, https://github.com/twisted/twisted/pull/1690,  for https://github.com/twisted/twisted/issues/10301 which relates to this stuff and was merged on 2022-05-10. But there hasn't been a Twisted release since 2022-04-11 so none of the versions include right now.

So now I want to try to build it from source and use it in Synapse to see if that fixes the problem. How to do? ⏩

Reply:
> So now I am curious and want to try to build Twisted from source and use it in Synapse to see if that fixes the problem. How to do? ⏩

I got it working with these steps:

 1. Delete [`Twisted` from `pyproject.toml`](https://github.com/matrix-org/synapse/blob/1c910e2216a2f7fa57971ffbdede509fd493ebb0/pyproject.toml#L122) so you don't run into `SolverProblemError` (`So, because matrix-synapse depends on both Twisted (>=18.9.0) and twisted (rev trunk), version solving failed.`)
 1. `poetry add git+https://github.com/twisted/twisted.git#trunk`
 
 And seems to fix all of the `Failed to detach context` errors 🎉 (good fix @richvdh 💪)

Reply:
`twisted==22.8.0` (2022-09-06) was released 🚀

Need to try it out